data/summary/types/summary.go

```diff
@@ -249,48 +264,123 @@ type Period interface {
+//		if currentPeriod.Equal(lastBucketPeriod) || currentPeriod.Before(lastBucketPeriod) {
+//
+//			gapPeriods := int(lastBucketPeriod.Sub(currentPeriod).Hours())
+//			if gapPeriods < len(*buckets) {
+//				if !(*buckets)[len(*buckets)-gapPeriods-1].Date.Equal(currentPeriod) {
+//					return errors.New("Potentially damaged buckets, offset jump did not find intended record.")
+//				}
+//				(*buckets)[len(*buckets)-gapPeriods-1] = newStat
+//				existingHour = true
+//			}
+//		}
+//
+//		// add hours for any gaps that this new bucket skipped
+//		statsGap := int(newStat.Date.Sub((*buckets)[len(*buckets)-1].Date).Hours())
+//		// only add gap buckets if the gap is shorter than max tracking amount
+//		if statsGap > 0 && statsGap < HoursAgoToKeep {
+//			gapBuckets := make(S, 0, statsGap)
+//			for i := statsGap; i > 1; i-- {
+//				newStatsTime := newStat.Date.Add(time.Duration(-i+1) * time.Hour)
+//				gapBuckets = append(gapBuckets, CreateBucket[A](newStatsTime))
+//			}
+//
+//			*buckets = append(*buckets, gapBuckets...)
+//		} else if statsGap > HoursAgoToKeep {
+//			// otherwise, the gap is larger than our tracking, delete all the old buckets for a clean state
+//			*buckets = make(S, 0, 1)
+//		}
+//	}
+//
+//	if existingHour == false {
+//		*buckets = append(*buckets, newStat)
+//	}
+//
+//	// remove extra hours to cap at X hours of buckets
+//	if len(*buckets) > HoursAgoToKeep {
+//		// zero out any to-be-trimmed buckets to lower their impact until reallocation
+//		for i := 0; i < len(*buckets)-HoursAgoToKeep; i++ {
+//			(*buckets)[i] = nil
+//		}
+//		*buckets = (*buckets)[len(*buckets)-HoursAgoToKeep:]
+//	}
+//
+//	return nil
+//}
+
+func AddBin[T BucketData, A BucketDataPt[T], S Buckets[T, A]](buckets *S, newBucket *Bucket[A, T]) error {
+	if len(*buckets) == 0 {
+		*buckets = append(*buckets, newBucket)
+		return nil
+	}
 
-			gapPeriods := int(lastBucketPeriod.Sub(currentPeriod).Hours())
-			if gapPeriods < len(*buckets) {
-				if !(*buckets)[len(*buckets)-gapPeriods-1].Date.Equal(currentPeriod) {
-					return errors.New("Potentially damaged buckets, offset jump did not find intended record.")
-				}
-				(*buckets)[len(*buckets)-gapPeriods-1] = newStat
-				existingHour = true
-			}
+	lastBucketPeriod := (*buckets)[len(*buckets)-1].Date
+	firstBucketPeriod := (*buckets)[0].Date
+	newPeriod := newBucket.Date
+	statsGap := 0
+	var gapStart time.Time
+	var gapEnd time.Time
+
+	// get the gap range
+	if newPeriod.After(lastBucketPeriod) {
+		// check if we are so far ahead that we should just nuke it
+		if newPeriod.Add(-HoursAgoToKeep * time.Hour).After(lastBucketPeriod) {
+			*buckets = make(S, 0, 1)
+		} else {
+			statsGap = int(newPeriod.Sub(lastBucketPeriod).Hours())
+			gapStart = lastBucketPeriod.Add(time.Hour)
+			gapEnd = newPeriod
+		}
+	} else if newPeriod.Before(firstBucketPeriod) {
+		// check if we are so far behind that we are irrelevant
+		if newPeriod.Before(lastBucketPeriod.Add(HoursAgoToKeep * -time.Hour)) {
```

Comment:
this is as intended, we check from the end in case its still in range of the latest data, we always check from the latest, as we track 60d from the newest, not the oldest.